pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,518 @@
+func (h *Handler) checkLonghornVolumeDetach() (bool, error) {
+	volumes, err := h.longhornVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		logrus.Warnf("volume error %v", err)
+		return false, err
+	}
+
+	for _, volume := range volumes {
+		logrus.Infof("volume state: %v", volume.Status.State)
+		if volume.Status.State != "detached" {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (h *Handler) checkPrometheusStatusAndStart() bool {
+	// check prometheus cattle-monitoring-system/rancher-monitoring-prometheus replica
+	prometheus, err := h.prometheusCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringPrometheus)
+	if err != nil {
+		logrus.Warnf("prometheus get error %v", err)
+		return false
+	}
+
+	logrus.Infof("prometheus: %v", *prometheus.Spec.Replicas)
+	// check started or not
+	if *prometheus.Spec.Replicas == 0 {
+		logrus.Infof("start prometheus")
+		prometheusCopy := prometheus.DeepCopy()
+		if replicas, err := strconv.Atoi(prometheus.Annotations[ReplicaStorageNetworkAnnotation]); err == nil {
+			*prometheusCopy.Spec.Replicas = int32(replicas)
+		}
+		delete(prometheusCopy.Annotations, ReplicaStorageNetworkAnnotation)
+
+		if _, err := h.prometheus.Update(prometheusCopy); err != nil {
+			logrus.Warnf("prometheus update error %v", err)
+			return false
+		}
+		return false
+	}
+
+	return true
+}
+
+func (h *Handler) checkAltermanagerStatusAndStart() bool {
+	// check alertmanager cattle-monitoring-system/rancher-monitoring-alertmanager replica
+	alertmanager, err := h.alertmanagerCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringAlertmanager)
+	if err != nil {
+		logrus.Warnf("alertmanager get error %v", err)
+		return false
+	}
+
+	logrus.Infof("alertmanager: %v", *alertmanager.Spec.Replicas)
+	// check started or not
+	if *alertmanager.Spec.Replicas == 0 {
+		logrus.Infof("start alertmanager")
+		alertmanagerCopy := alertmanager.DeepCopy()
+		if replicas, err := strconv.Atoi(alertmanager.Annotations[ReplicaStorageNetworkAnnotation]); err == nil {
+			*alertmanagerCopy.Spec.Replicas = int32(replicas)
+		}
+		delete(alertmanagerCopy.Annotations, ReplicaStorageNetworkAnnotation)
+
+		if _, err := h.alertmanager.Update(alertmanagerCopy); err != nil {
+			logrus.Warnf("alertmanager update error %v", err)
+			return false
+		}
+		return false
+	}
+
+	return true
+}
+
+func (h *Handler) checkGrafanaStatusAndStart() bool {
+	// check deployment cattle-monitoring-system/rancher-monitoring-grafana replica
+	grafana, err := h.deploymentCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringGrafana)
+	if err != nil {
+		logrus.Warnf("grafana get error %v", err)
+		return false
+	}
+
+	logrus.Infof("Grafana: %v", *grafana.Spec.Replicas)
+	// check started or not
+	if *grafana.Spec.Replicas == 0 {
+		logrus.Infof("start grafana")
+		grafanaCopy := grafana.DeepCopy()
+		if replicas, err := strconv.Atoi(grafana.Annotations[ReplicaStorageNetworkAnnotation]); err == nil {
```

Comment:
It looks like we don't handle `err != nil` here.

Reply:
I set default value to handle and ignore error